Further reduce use of memcpy()

cdumez opened the pull request:
#### 970bbbc53b2d8dd3be98c96099298d221607f774
<pre>
Further reduce use of memcpy()
<a href="https://bugs.webkit.org/show_bug.cgi?id=284897">https://bugs.webkit.org/show_bug.cgi?id=284897</a>

Reviewed by Darin Adler.

* Source/WebCore/Modules/webaudio/WaveShaperDSPKernel.cpp:
(WebCore::WaveShaperDSPKernel::process):
(WebCore::WaveShaperDSPKernel::processCurve):
(WebCore::WaveShaperDSPKernel::processCurve2x):
(WebCore::WaveShaperDSPKernel::processCurve4x):
* Source/WebCore/Modules/webaudio/WaveShaperDSPKernel.h:
* Source/WebCore/Modules/webaudio/WaveShaperProcessor.cpp:
(WebCore::WaveShaperProcessor::process):
* Source/WebCore/platform/audio/DownSampler.cpp:
(WebCore::DownSampler::process):
* Source/WebCore/platform/audio/DownSampler.h:
* Source/WebCore/platform/audio/UpSampler.cpp:
(WebCore::UpSampler::process):
* Source/WebCore/platform/audio/UpSampler.h:
* Source/WebCore/platform/audio/cocoa/AudioSampleBufferList.cpp:
(WebCore::AudioSampleBufferList::copyFrom):
(WebCore::AudioSampleBufferList::copyTo):
* Source/WebCore/platform/audio/cocoa/CARingBuffer.cpp:
(WebCore::FetchABL):
* Source/WebCore/platform/cocoa/SharedVideoFrameInfo.h:
* Source/WebCore/platform/cocoa/SharedVideoFrameInfo.mm:
(WebCore::SharedVideoFrameInfo::encode):
(WebCore::SharedVideoFrameInfo::writePixelBuffer):
(WebCore::SharedVideoFrameInfo::writeVideoFrameBuffer):
* Source/WebCore/platform/graphics/WidthCache.h:
(WebCore::WidthCache::SmallStringKey::SmallStringKey):
(WebCore::WidthCache::SmallStringKey::copySmallCharacters):
* Source/WebCore/platform/graphics/cocoa/ANGLEUtilitiesCocoa.mm:
(WebCore::newRasterizationRateMap):
* Source/WebCore/platform/mediastream/mac/MockAudioSharedUnit.mm:
(WebCore::MockAudioSharedInternalUnit::generateSampleBuffers):
(WebCore::MockAudioSharedInternalUnit::render):
* Source/WebCore/platform/video-codecs/cocoa/RTCVideoDecoderVTBAV1.mm:
(computeAV1InputFormat):
* Source/WebCore/platform/xr/openxr/OpenXRInstance.cpp:
(PlatformXR::Instance::Impl::Impl):
* Source/WebCore/xml/parser/XMLDocumentParserLibxml2.cpp:
(WebCore::OffsetBuffer::readOutBytes):
(WebCore::readFunc):
(WebCore::XMLParserContext::createMemoryParser):
* Source/WebKit/WebProcess/GPU/webrtc/SharedVideoFrame.cpp:
(WebKit::SharedVideoFrameWriter::writeBuffer):
* Source/WebKitLegacy/mac/History/BinaryPropertyList.cpp:
(BinaryPropertyListSerializer::BinaryPropertyListSerializer):
(BinaryPropertyListSerializer::writeArrayStart):
(BinaryPropertyListSerializer::writeArrayEnd):
(BinaryPropertyListSerializer::writeDictionaryStart):
(BinaryPropertyListSerializer::writeDictionaryEnd):
(BinaryPropertyListSerializer::startObject):
* Source/WebKitLegacy/mac/Plugins/WebBasePluginPackage.mm:
(-[WebBasePluginPackage isNativeLibraryData:]):
* Tools/TestWebKitAPI/Tests/WTF/cocoa/URLExtras.mm:
(TestWebKitAPI::dataAsString):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebPushDaemon.mm:
(TestWebKitAPI::defaultWebPushDaemonConfiguration):

Canonical link: <a href="https://commits.webkit.org/288121@main">https://commits.webkit.org/288121@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1aa1eadbd99ebb26d7c4b79fd7e536eaf8221f05

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/81889 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/1416 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/35845 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/86446 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/32891 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/83995 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/1450 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/9239 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/63872 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/21599 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/84959 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/1072 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/74524 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/44156 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/972 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/28701 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/31340 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/72317 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/29310 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/87877 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/9131 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/6509 "Found 1 new test failure: http/wpt/mediarecorder/record-96KHz-sources.html (failure)") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/72231 "Found 2 new test failures: imported/w3c/web-platform-tests/css/css-view-transitions/content-with-transparent-background.html imported/w3c/web-platform-tests/css/css-view-transitions/sibling-frames-transition.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/9316 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/70343 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/71454 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/15548 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/14473 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/512 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12697 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/9082 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/14618 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/8922 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/12447 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/10730 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/platform/audio/FFTConvolver.cpp:
```diff
@@ -64,28 +64,30 @@ void FFTConvolver::process(FFTFrame* fftKernel, const float* sourceP, float* des
             return;
 
-        memcpy(inputP + m_readWriteIndex, sourceP, sizeof(float) * divisionSize);
+        IGNORE_WARNINGS_BEGIN("restrict")
```
Very surprising that we both get a warning here and that we need to ignore it.

cdumez replied:
Only on the WPE bot.

darinadler commented on Source/WebCore/platform/audio/UpSampler.h:
```diff
@@ -42,8 +42,8 @@ class UpSampler final {
 
-    // The destination buffer |destP| is of size sourceFramesToProcess * 2.
-    void process(const float* sourceP, float* destP, size_t sourceFramesToProcess);
+    // The destination buffer |destination| is of size sourceFramesToProcess * 2.
```
Why pass the size twice, then?

cdumez replied:
Ok. I'll look into not passing the size and making sure that `source` has `size() == sourceFramesToProcess`

darinadler commented on Source/WebCore/platform/cocoa/SharedVideoFrameInfo.mm:
```diff
@@ -298,12 +299,12 @@
     encode(data);
-    data += sizeof(SharedVideoFrameInfo);
-    return webrtc::copyVideoFrameBuffer(frameBuffer, data);
+    data = data.subspan(sizeof(SharedVideoFrameInfo));
```
Seems like we should not reassign `data`, just put this expression into the next line function call.

darinadler commented on Source/WebCore/platform/graphics/WidthCache.h:
```diff
@@ -83,13 +84,13 @@ class WidthCache {
             else {
-                for (auto character : source)
-                    *destination++ = character;
+                for (auto [i, character] : indexedRange(source))
```
Can this use `zippedRange` instead?

cdumez replied:
I wasn't sure if `zippedRange` supported mutation (I need to assign the character here). I'll give it a try and see if it works.

darinadler commented on Source/WebKitLegacy/mac/History/BinaryPropertyList.cpp:
```diff
@@ -498,28 +498,28 @@ BinaryPropertyListSerializer::BinaryPropertyListSerializer(BinaryPropertyListWri
-    storeLength(trailer + 8, m_plan.objectCount());
-    storeLength(trailer + 16, m_plan.objectCount() - 1);
-    storeLength(trailer + 24, m_offsetTableStart);
+    storeLength(trailer.subspan(8).data(), m_plan.objectCount());
```
Why not have `storeLength` take a `span`?

darinadler commented on Source/WebKitLegacy/mac/History/BinaryPropertyList.cpp:
```diff
@@ -585,14 +585,14 @@ void BinaryPropertyListSerializer::writeUniqueString(const String& string)
 size_t BinaryPropertyListSerializer::writeArrayStart()
 {
-    return m_currentAggregateBufferByte - m_buffer;
+    return m_currentAggregateBufferByte - m_buffer.data();
```
Why not change both data members to spans?

cdumez replied:
I've updated the 2 raw pointer data members (including m_currentAggregateBufferByte) to be indexes in m_buffer instead of raw pointers.

cdumez committed:
Further reduce use of memcpy()
https://bugs.webkit.org/show_bug.cgi?id=284897

Reviewed by Darin Adler.

* Source/WebCore/Modules/webaudio/WaveShaperDSPKernel.cpp:
(WebCore::WaveShaperDSPKernel::process):
(WebCore::WaveShaperDSPKernel::processCurve):
(WebCore::WaveShaperDSPKernel::processCurve2x):
(WebCore::WaveShaperDSPKernel::processCurve4x):
* Source/WebCore/Modules/webaudio/WaveShaperDSPKernel.h:
* Source/WebCore/Modules/webaudio/WaveShaperProcessor.cpp:
(WebCore::WaveShaperProcessor::process):
* Source/WebCore/platform/audio/DownSampler.cpp:
(WebCore::DownSampler::process):
* Source/WebCore/platform/audio/DownSampler.h:
* Source/WebCore/platform/audio/UpSampler.cpp:
(WebCore::UpSampler::process):
* Source/WebCore/platform/audio/UpSampler.h:
* Source/WebCore/platform/audio/cocoa/AudioSampleBufferList.cpp:
(WebCore::AudioSampleBufferList::copyFrom):
(WebCore::AudioSampleBufferList::copyTo):
* Source/WebCore/platform/audio/cocoa/CARingBuffer.cpp:
(WebCore::FetchABL):
* Source/WebCore/platform/cocoa/SharedVideoFrameInfo.h:
* Source/WebCore/platform/cocoa/SharedVideoFrameInfo.mm:
(WebCore::SharedVideoFrameInfo::encode):
(WebCore::SharedVideoFrameInfo::writePixelBuffer):
(WebCore::SharedVideoFrameInfo::writeVideoFrameBuffer):
* Source/WebCore/platform/graphics/WidthCache.h:
(WebCore::WidthCache::SmallStringKey::SmallStringKey):
(WebCore::WidthCache::SmallStringKey::copySmallCharacters):
* Source/WebCore/platform/graphics/cocoa/ANGLEUtilitiesCocoa.mm:
(WebCore::newRasterizationRateMap):
* Source/WebCore/platform/mediastream/mac/MockAudioSharedUnit.mm:
(WebCore::MockAudioSharedInternalUnit::generateSampleBuffers):
(WebCore::MockAudioSharedInternalUnit::render):
* Source/WebCore/platform/video-codecs/cocoa/RTCVideoDecoderVTBAV1.mm:
(computeAV1InputFormat):
* Source/WebCore/platform/xr/openxr/OpenXRInstance.cpp:
(PlatformXR::Instance::Impl::Impl):
* Source/WebCore/xml/parser/XMLDocumentParserLibxml2.cpp:
(WebCore::OffsetBuffer::readOutBytes):
(WebCore::readFunc):
(WebCore::XMLParserContext::createMemoryParser):
* Source/WebKit/WebProcess/GPU/webrtc/SharedVideoFrame.cpp:
(WebKit::SharedVideoFrameWriter::writeBuffer):
* Source/WebKitLegacy/mac/History/BinaryPropertyList.cpp:
(BinaryPropertyListSerializer::BinaryPropertyListSerializer):
(BinaryPropertyListSerializer::writeArrayStart):
(BinaryPropertyListSerializer::writeArrayEnd):
(BinaryPropertyListSerializer::writeDictionaryStart):
(BinaryPropertyListSerializer::writeDictionaryEnd):
(BinaryPropertyListSerializer::startObject):
* Source/WebKitLegacy/mac/Plugins/WebBasePluginPackage.mm:
(-[WebBasePluginPackage isNativeLibraryData:]):
* Tools/TestWebKitAPI/Tests/WTF/cocoa/URLExtras.mm:
(TestWebKitAPI::dataAsString):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebPushDaemon.mm:
(TestWebKitAPI::defaultWebPushDaemonConfiguration):

Canonical link: https://commits.webkit.org/288121@main
